src/box/memtx_engine.cc

```diff
@@ -926,10 +923,17 @@ memtx_engine_bootstrap(struct engine *engine)
 		return -1;
 
 	int rc;
-	struct xrow_header row;
+	struct recovery_request request;
 	enum snapshot_recovery_state state = SNAPSHOT_RECOVERY_NOT_STARTED;
-	while ((rc = xlog_cursor_next(&cursor, &row, true)) == 0) {
-		rc = memtx_engine_recover_snapshot_row(memtx, &row, &state);
+	while ((rc = xlog_cursor_next(&cursor, &request.header, false)) == 0) {
+		rc = recovery_request_decode(&request.header, &request);
```

Comment:
I don't like that we have to use `recovery_request_decode` here. It'd be nice if there was a synchronous xlog reader defined in `xlog.h` and the asynchronous reader would only implement inter-thread communication over it.